packages/api/src/agent.ts

```diff
@@ -0,0 +1,298 @@
+        identifier: opts.identifier,
+        password: opts.password,
+      })
+      this.session = {
+        accessJwt: res.data.accessJwt,
+        refreshJwt: res.data.refreshJwt,
+        handle: res.data.handle,
+        did: res.data.did,
+      }
+      return res
+    } finally {
+      if (this.session) {
+        this._persistSession?.('create', this.session)
+      } else {
+        this._persistSession?.('create-failed', undefined)
+      }
+    }
+  }
+
+  /**
+   * Resume a pre-existing session with this agent.
+   */
+  async resumeSession(
+    session: AtpSessionData,
+  ): Promise<ComAtprotoSessionGet.Response> {
+    this.session = session
+    try {
+      const res = await this.api.com.atproto.session.get()
+      if (!res.success || res.data.did !== this.session.did) {
+        throw new Error('Invalid session')
+      }
+      return res
+    } catch (e: any) {
+      this.session = undefined
+      throw e
+    } finally {
+      if (this.session) {
+        this._persistSession?.('create', this.session)
+      } else {
+        this._persistSession?.('create-failed', undefined)
+      }
+    }
+  }
+
+  /**
+   * Internal helper to add authorization headers to requests.
+   */
+  private _addAuthHeader(reqHeaders: Record<string, string>) {
+    if (!reqHeaders.authorization && this.session?.accessJwt) {
+      return {
+        ...reqHeaders,
+        authorization: `Bearer ${this.session.accessJwt}`,
+      }
+    }
+    return reqHeaders
+  }
+
+  /**
+   * Internal fetch handler which adds access-token management
+   */
+  private async _fetch(
+    reqUri: string,
+    reqMethod: string,
+    reqHeaders: Record<string, string>,
+    reqBody: any,
+  ): Promise<AtpAgentFetchHandlerResponse> {
+    if (!AtpAgent.fetch) {
+      throw new Error('AtpAgent fetch() method not configured')
+    }
+
+    // wait for any active session-refreshes to finish
+    await this._refreshSessionPromise
+
+    // send the request
+    let res = await AtpAgent.fetch(
+      reqUri,
+      reqMethod,
+      this._addAuthHeader(reqHeaders),
+      reqBody,
+    )
+
+    // handle session-refreshes as needed
+    if (isErrorResponse(res, ['ExpiredToken']) && this.session?.refreshJwt) {
+      // attempt refresh
+      await this._refreshSession()
+
+      // resend the request with the new access token
+      res = await AtpAgent.fetch(
+        reqUri,
+        reqMethod,
+        this._addAuthHeader(reqHeaders),
+        reqBody,
+      )
+    }
+
+    return res
+  }
+
+  /**
+   * Internal helper to refresh sessions
+   * - Wraps the actual implementation in a promise-guard to ensure only
+   *   one refresh is attempted at a time.
+   */
+  private async _refreshSession() {
+    if (this._refreshSessionPromise) {
+      return this._refreshSessionPromise
+    }
+    this._refreshSessionPromise = this._refreshSessionInner()
+    await this._refreshSessionPromise
+    this._refreshSessionPromise = undefined
```

Comment:
Do we want to clear this even if `this._refreshSessionPromise` rejects?

Reply:
Bleh yes of course. Aaaaand now I'm realizing I've got that mistake in multiple places in the app repo! Hooray!